src/service_daemon.rs

```diff
@@ -582,32 +584,69 @@ impl ServiceDaemon {
-    let test_packet = DnsOutgoing::new(0).to_packet_data();
-    sock.send_to(&test_packet, &multicast_addr)
-        .map_err(|e| e_fmt!("send multicast packet on addr {}: {}", intf_ip, e))?;
-
-    Ok(sock)
+    // Such socket has to bind to INADDR_ANY or IN6ADDR_ANY.
+    let intf_ip = &intf.ip();
+    match intf_ip {
+        IpAddr::V4(ip) => {
+            let addr = SocketAddrV4::new(Ipv4Addr::new(0, 0, 0, 0), MDNS_PORT);
+            let sock = new_socket(addr.into(), true)?;
+
+            // Join mDNS group to receive packets.
+            sock.join_multicast_v4(&GROUP_ADDR_V4, ip)
+                .map_err(|e| e_fmt!("join multicast group on addr {}: {}", intf_ip, e))?;
+
+            // Set IP_MULTICAST_IF to send packets.
+            sock.set_multicast_if_v4(ip)
+                .map_err(|e| e_fmt!("set multicast_if on addr {}: {}", ip, e))?;
+
+            // Test if we can send packets successfully.
+            let multicast_addr = SocketAddrV4::new(GROUP_ADDR_V4, MDNS_PORT).into();
+            let test_packet = DnsOutgoing::new(0).to_packet_data();
+            sock.send_to(&test_packet, &multicast_addr)
+                .map_err(|e| e_fmt!("send multicast packet on addr {}: {}", ip, e))?;
+            Ok(sock)
+        }
+        IpAddr::V6(ip) => {
+            let mut addr = SocketAddrV6::new(
+                Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 0),
+                MDNS_PORT,
+                0,
+                0,
+            );
+            addr.set_scope_id(intf.index.unwrap_or(0));
+            let sock = new_socket(addr.into(), true)?;
+
+            // Join mDNS group to receive packets.
+            sock.join_multicast_v6(&GROUP_ADDR_V6, intf.index.unwrap_or(0))
+                .map_err(|e| e_fmt!("join multicast group on addr {}: {}", ip, e))?;
+
+            // Set IP6_MULTICAST_IF to send packets.
+            sock.set_multicast_if_v6(intf.index.unwrap_or(0))
+                .map_err(|e| e_fmt!("set multicast_if on addr {}: {}", ip, e))?;
+
+            // Test if we can send packets successfully.
+            let multicast_addr = SocketAddrV6::new(GROUP_ADDR_V6, MDNS_PORT, 0, 0).into();
+            let test_packet = DnsOutgoing::new(0).to_packet_data();
+            sock.send_to(&test_packet, &multicast_addr)
+                .map_err(|e| e_fmt!("send multicast packet on addr {}: {}", ip, e))?;
+            Ok(sock)
+        }
+    }
 }
 
 /// Creates a new UDP socket to bind to `port` with REUSEPORT option.
 /// `non_block` indicates whether to set O_NONBLOCK for the socket.
-fn new_socket(ipv4: Ipv4Addr, port: u16, non_block: bool) -> Result<Socket> {
-    let fd = Socket::new(socket2::Domain::IPV4, socket2::Type::DGRAM, None)
+fn new_socket(addr: SocketAddr, non_block: bool) -> Result<Socket> {
+    let domain = if addr.is_ipv6() {
+        socket2::Domain::IPV6
+    } else if addr.is_ipv4() {
+        socket2::Domain::IPV4
+    } else {
+        panic!()
+    };
```

Comment:
Wouldn't a `match &addr` also work here, and can avoid `panic!` ?

Reply:
Yes, its much better with a match!